src/chainstate/burn/operations/leader_block_commit.rs

```diff
@@ -905,12 +912,86 @@ impl LeaderBlockCommitOp {
         // epoch marker field -- for example, to signal support for a new epoch or to be
         // forwards-compatible with it -- but cannot put a lesser number in.
         /////////////////////////////////////////////////////////////////////////////////////
+        self.check_epoch_commit(epoch_id)?;
+        Ok(())
+    }
+
+    pub fn check(
+        &self,
+        burnchain: &Burnchain,
+        tx: &mut SortitionHandleTx,
+        reward_set_info: Option<&RewardSetInfo>,
+    ) -> Result<(), op_error> {
+        /////////////////////////////////////////////////////////////////////////////////////
+        // There must be a burn
+        /////////////////////////////////////////////////////////////////////////////////////
+
+        let apparent_sender_address = self
+            .apparent_sender
+            .to_bitcoin_address(BitcoinNetworkType::Mainnet);
+
+        if self.burn_fee == 0 {
+            warn!("Invalid block commit: no burn amount";
+                  "apparent_sender" => %apparent_sender_address
+            );
+            return Err(op_error::BlockCommitBadInput);
+        }
         let epoch = SortitionDB::get_stacks_epoch(tx, self.block_height)?.expect(&format!(
             "FATAL: impossible block height: no epoch defined for {}",
             self.block_height
         ));
 
-        self.check_epoch_commit(epoch.epoch_id)?;
+        let intended_modulus = (self.burn_block_mined_at() + 1) % BURN_BLOCK_MINED_AT_MODULUS;
+        let actual_modulus = self.block_height % BURN_BLOCK_MINED_AT_MODULUS;
```

Comment:
I think this is a dumb question, but I'm having trouble understanding how we can just use the modulus here. Where are we sure that we're in the right window of `BURN_BLOCK_MINED_AT_MODULUS`?

Reply:
The window is always the last 5 burnchain transactions sent by this miner, plus the current one (each tx must consume a UTXO produced by the previous).  If a block-commit's modulus is valid, then at most one of the following is true:

* The block-commit landed in the intended burnchain block

* The block-commit is so late that it's no longer part of the window, and your sortition weight is now 0.

This all happens in `make_min_median_distribution()` in `src/chainstate/burn/distribution.rs`.  The call site for this function is in `BurnchainStateTransition::from_block_ops()`.